modules/transport-netty4/src/test/java/org/elasticsearch/transport/netty4/Netty4WriteThrottlingHandlerTests.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+package org.elasticsearch.transport.netty4;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelOutboundHandlerAdapter;
+import io.netty.channel.ChannelPromise;
+import io.netty.channel.embedded.EmbeddedChannel;
+
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+import org.elasticsearch.test.ESTestCase;
+import org.junit.After;
+import org.junit.Before;
+
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.ExecutionException;
+
+import static org.hamcrest.Matchers.hasSize;
+import static org.hamcrest.Matchers.instanceOf;
+import static org.hamcrest.Matchers.lessThan;
+import static org.hamcrest.Matchers.lessThanOrEqualTo;
+
+public class Netty4WriteThrottlingHandlerTests extends ESTestCase {
+
+    private SharedGroupFactory.SharedGroup transportGroup;
+
+    @Before
+    public void createGroup() {
+        final SharedGroupFactory sharedGroupFactory = new SharedGroupFactory(Settings.EMPTY);
+        transportGroup = sharedGroupFactory.getTransportGroup();
+    }
+
+    @After
+    public void stopGroup() {
+        transportGroup.shutdown();
+    }
+
+    public void testThrottlesLargeMessage() throws ExecutionException, InterruptedException {
+        final List<ByteBuf> seen = new CopyOnWriteArrayList<>();
+        final EmbeddedChannel embeddedChannel = new EmbeddedChannel(
+            new CapturingHandler(seen),
+            new Netty4WriteThrottlingHandler(new ThreadContext(Settings.EMPTY))
+        );
+        final int writeableBytes = Math.toIntExact(embeddedChannel.bytesBeforeUnwritable());
+        assertThat(writeableBytes, lessThan(Netty4WriteThrottlingHandler.MAX_BYTES_PER_WRITE));
```

Comment:
This is not an assumption, we hard code this constant as the maximum message size we send in one go. When I introduced this in 2019 it was a best-guess for performance, it's not directly coming from Netty and rather a best guess regarding the tradeoff between latency, memory and overhead for handling individual chunks.

Reply:
I was thinking about it the other way around. An assumption of this test is that the buffer size in netty is smaller than the buffer size we picked, since otherwise the calculations in the test will not work (if netty allowed 2x our buffer size, we would for instance see 2 byte-bufs on the write).

Reply:
Ah got it :) added a comment to that effect